src/printer.js

```diff
@@ -2032,27 +2032,68 @@ function printJSXElement(path, options, print) {
     var children = [];
 
     path.map(
-      function(childPath) {
+      function(childPath, i) {
         var child = childPath.getValue();
+        let lastIdx = children.length - 1;
+        let lastChild = children[lastIdx];
 
         if (
           namedTypes.Literal.check(child) && typeof child.value === "string"
         ) {
           if (/\S/.test(child.value)) {
             const beginBreak = child.value.match(/^\s*\n/);
             const endBreak = child.value.match(/\n\s*$/);
+            const beginSpace = child.value.match(/^\s+/);
+            const endSpace = child.value.match(/\s+$/);
+
+            if (beginBreak) {
+              children.push(hardline);
+            } else if (beginSpace) {
+              let prev = util.getLast(children);
+              if (prev && prev.type === "line") {
+                // TODO ifBreak with else
+                children.push("{' '}", softline);
+              } else {
+                children.push(" ");
+              }
+            }
+
+            children.push(child.value.replace(/^\s+|\s+$/g, ""));
+
+            if (endBreak) {
+              children.push(hardline);
+            } else if (endSpace) {
+              children.push(" ");
+            }
 
-            children.push(
-              beginBreak ? hardline : "",
-              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
-              endBreak ? hardline : ""
-            );
           } else if (/\n/.test(child.value)) {
```

Comment:
... which makes me realize, we need to add a `literalline` for each `\n` found, not just a single `hardline`, I think... will add tests for that.